### PR TITLE
Tickets/dm 35736

### PR DESCRIPTION
**Pull request by pcortes80**
- Create **src** and **dependencies** directories a placeholders.
- Create **README.md** file.
- Create **versionHistory.md** file.
- Create **doc** directories with more .md files.

### CODE REVIEW

File: doc/versionHistory.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Version History
+
+0.0.1
+
+- Add README.md, versionHistory.md, moduleRelays.md, moduleDaq, and moduleIlc.md.
```
Comment (teweitsai) [inline doc/versionHistory.md]: Add `README.md`, `versionHistory.md`, `moduleRelays.md`, `moduleDaq.md`, and `moduleIlc.md`.
The final two are the placeholders at this moment.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,76 @@
-# ts_m2fpga
-M2 Controller FPGA code
+# M2 Cell Controller FPGA Code
+
+## Overview
+
+This project will improve the FPGA code in the [M2 Cell Controller LaVIEW project](https://github.com/lsst-ts/ts_mtm2_cell).
```
Comment (teweitsai) [inline README.md]: LaVIEW --> LabVIEW

File: README.md [hunk 3]
```diff
@@ -1,2 +1,76 @@
...
+In that version, the VI named **4-port serial Master Slave.vi**, under `Chassis (cRIO-9038) >> FPGA Target (RIO0, cRIO-9038)` in the Project Explorer, interacts with the hardware inputs and outputs.
+
+## Directory Structure
+
+```bash
```
Comment (teweitsai) [inline README.md]: Not "bash" but "text".

File: README.md [hunk 4]
```diff
@@ -1,2 +1,76 @@
...
+ts_m2fpga
+├── src
+│   ├── m2fpga.lvproj
+│   ├── mainFPGA.vi
+│   ├── moduleRelays.vi
+│   ├── moduleDaq.vi
+│   ├── moduleIlc.vi
```
Comment (teweitsai) [inline README.md:17]: It is weird to me that you have the vi named begin from `module`. It is a reserved word with the specific meaning at the most of time. Maybe `commRelays.vi`, `commDaq.vi`, and `commIlc.vi`?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,76 @@
...
+├── doc
+│   ├── versionHistory.md
+│   ├── moduleRelays.md
+│   ├── moduleDaq.md
+│   ├── moduleIlc.md
+├── dependencies
```
Comment (teweitsai) [inline README.md]: The `dependencies` is a reserved word to say this module depends on "what" libraries, modules, etc. Here, they are the output from the LabVIEW code generator. They are the dependencies for `ts_m2cellcpp` but not this module. Consider rename to `fpgaCode` or something similar.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,76 @@
...
+│   ├── NiFpga_mainFPGA.lvbitx
+│   ├── NiFpga.c
+│   ├── NiFpga.h
+├── README.md
+└── .gitignore
+```
+
+- `src`: in here are all LabVIEW files.
+- `doc`: documentation like `versionHistory.md`.
+- `dependencies`: in here are all files generated by the FPGA Interface C API.
```
Comment (teweitsai) [inline README.md]: See above comment for `dependencies`.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,76 @@
...
+├── dependencies
+│   ├── NiFpga_mainFPGA.h
+│   ├── NiFpga_mainFPGA.lvbitx
+│   ├── NiFpga.c
+│   ├── NiFpga.h
+├── README.md
+└── .gitignore
+```
+
+- `src`: in here are all LabVIEW files.
```
Comment (teweitsai) [inline README.md]: `src`: contains the LabVIEW FPGA project.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,76 @@
-# ts_m2fpga
-M2 Controller FPGA code
+# M2 Cell Controller FPGA Code
```
Comment (teweitsai) [inline README.md]: Give the full name of "FPGA" before the abbreviation.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,76 @@
...
+└── .gitignore
+```
+
+- `src`: in here are all LabVIEW files.
+- `doc`: documentation like `versionHistory.md`.
+- `dependencies`: in here are all files generated by the FPGA Interface C API.
+
+## FPGA Interface C API
+
+The FPGA Interface C API enables C/C++ applications to interact directly with the compiled LabVIEW FPGA VI on CompactRIO devices, generating C functions calls to read from and write to named controls and indicators, perform DMA data transfers, and wait on and acknowledge interrupts.
```
Comment (teweitsai) [inline README.md]: on CompactRIO devices --> on real-time controllers

Comment (teweitsai) [reply README.md]: Give the full name of "DMA" before the abbreviation.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,76 @@
...
+
+```
+NiFpga_<FPGA_VI_name>.h
+```
+
+For instance, if the FPGA VI name is `mainFPGA.vi`, the header file will be named `NiFpga_mainFPGA.h`.
+
+## FPGA Modules
+
+The LabVIEW FPGA Code will have three software modules running in parallel: (1) moduleRelays, (2) moduleDaq, and (3) moduleIlc.
```
Comment (teweitsai) [inline README.md]: See my comments of vi names above.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,76 @@
...
+This module will be in `src/` directory.
+Follow [here](doc/moduleRelays.md) for details.
+
+### Module DAQ
+
+This module will be explained in this [JIRA ticket](https://jira.lsstcorp.org/browse/DM-35829).
+
+### Module ILC
+
+This module will be explained in this [JIRA ticket](https://jira.lsstcorp.org/browse/DM-35830).
```
Comment (teweitsai) [inline README.md]: See my review of "module" names above.

File: doc/moduleRelays.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+The `controlRelays` is an 8-bit unsigned (U8) interger control.
+It is declared as a typedef enum like this:
+
+```
+{
+   NiFpga_mainFPGA_ControlU8_controlRelays = 0x18002,
+} NiFpga_mainFPGA_ControlU8;
+```
+Do not care about the 0x18002 value.
```
Comment (teweitsai) [inline doc/moduleRelays.md]: A new line from the above "```".

File: doc/moduleRelays.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+```c 
+NiFpga_WriteU8(session, NiFpga_mainFPGA_ControlU8_controlRelays, 0);
+```
+
+where `session` handles a currently FPGA session, opened with:
+
+```c
+NiFpga_Session session;
+```
+and `0` is the `uint8_t` value wrote to the control.
```
Comment (teweitsai) [inline doc/moduleRelays.md]: A new line from the above "```".

Comment (teweitsai) [reply doc/moduleRelays.md]: wrote --> written

File: doc/moduleRelays.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+uint8_t value = 0; /* Declare the value variable */
+NiFpga_ReadU8(session, NiFpga_mainFPGA_IndicatorU8_debugControlRelays = 0x18006, &value);
+```
+
+where `session` handles a currently session FPGA, opened with:
+
+```c
+NiFpga_Session session;
+```
+and `value` outputs what was read.
```
Comment (teweitsai) [inline doc/moduleRelays.md]: A new line from the above "```".

File: doc/moduleRelays.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+The following table describes the meaning of each bit in the `controlRelay` control.
+
+| Bit |  0  |  1   |
+|:---:|:-----:|:-----:|
+|  0  | ILC Motor Power is Off | ILC Motor Power is On |
+|  1  | ILC Communication Power is Off  | ILC Communication is On |
+|  2  | Disable cRIO Interlock | Enable cRIO Interlock |
+|  3  | Default value of Reset Power Motor Breakers | Reset Power Motor Breakers |
+|  4  | Default value of Reset Communication Power Breakers | Reset Communication Power Breakers |
```
Comment (teweitsai) [inline doc/moduleRelays.md]: After I put the "reset bit 3 and 4" to be 1, do I need to put them back to 0 in the next functional call? Describe this if you know what will happen. If not, add a note here to say the details will be clarified in a latter time.

File: doc/moduleRelays.md [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+## Control Relay Debugging
+
+ The `debugControlRelays` is an 8-bit unsigned (U8) integer indicator used to check the input values written in the `controlRelay` are really read by the FPGA code. It is declared as a typedef enum like this:
+
+```
+{
+   NiFpga_mainFPGA_IndicatorU8_debugControlRelays = 0x18006,
+} NiFpga_mainFPGA_IndicatorU8;
+```
+Do not care about the 0x18006 value.
```
Comment (teweitsai) [inline doc/moduleRelays.md]: A new line from the above "```".

**Review by teweitsai — APPROVED** [review]
Looks good. The C/C++ developer should have the idea how to use this document now. For the .`gitignore` file, please put the details like this:
https://github.com/lsst-ts/ts_labview_tcp_json/blob/develop/.gitignore

File: doc/versionHistory.md (hunk 1 above)
Comment (pcortes80) [reply doc/versionHistory.md]: Done.

File: README.md (hunk 2 above)
Comment (pcortes80) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (pcortes80) [reply README.md]: Done

File: doc/moduleRelays.md (hunk 12 above)
Comment (pcortes80) [reply doc/moduleRelays.md]: Done

File: doc/moduleRelays.md (hunk 13 above)
Comment (pcortes80) [reply doc/moduleRelays.md]: Done

File: doc/moduleRelays.md (hunk 15 above)
Comment (pcortes80) [reply doc/moduleRelays.md]: Done

File: doc/moduleRelays.md (hunk 16 above)
Comment (pcortes80) [reply doc/moduleRelays.md]: Done

File: doc/moduleRelays.md (hunk 14 above)
Comment (pcortes80) [reply doc/moduleRelays.md]: Done

File: README.md (hunk 7 above)
Comment (pcortes80) [reply README.md]: Done

File: README.md (hunk 8 above)
Comment (pcortes80) [reply README.md]: Done

File: README.md (hunk 9 above)
Comment (pcortes80) [reply README.md]: Done

File: README.md (hunk 4 above)
Comment (pcortes80) [reply README.md:17]: Changed: 
- `moduleRelays.vi` --> `switchRelays.vi`
- `moduleDaq.vi` --> `readDaq.vi`
- `moduleIlc.vi` --> `ilcCommunication.vi`

Comment (pcortes80) [reply README.md:17]: Done.

File: README.md (hunk 5 above)
Comment (pcortes80) [reply README.md]: Changed `dependencies` to `fpgaInterface` directory name.

File: README.md (hunk 6 above)
Comment (pcortes80) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (pcortes80) [reply README.md]: Done.

File: README.md (hunk 11 above)
Comment (pcortes80) [reply README.md]: Done.

**Pull request merged**